### ISSUE TITLE
Contextualising not working

### ISSUE THREAD

**Issue by jamesrhester** [lark-parser/lark#838]
**Describe the bug**

The contextual lexer is not provided with the correct list of possible tokens for the corresponding parser state, and can therefore match a token that is not expected.

**To Reproduce**

The following grammar (simplified from the one I'm actually using) describes a stream of words. These words may be whitespace-separated undelimited text, text between single quotes, and "tables" of the form { 'key' : value }, where the key must be quoted and all whitespace within the table is optional.

Grammar:
```
// Demonstrate fail to contextualise

WHITESPACE :  /([ \t\n\r](?!;))|[ \t]/
%ignore WHITESPACE

NODELIM : /[a-z:]+/
DELIM : /'[a-z:]+'/

start : data*
data : table | NODELIM | quoted_value
table : "{" table_entry* "}"
table_entry : quoted_value ":" data
quoted_value : DELIM
```

Using lark 0.11.1 I get the following (also fails on Lark 0.7):
```
In [15]: p = lark.Lark(gram,parser="lalr",lexer="contextual") 
In [16]: p.parse("one two {'three' :four}") 
UnexpectedToken: Unexpected token Token('NODELIM', ':four') at line 1, column 18.
Expected one of: 
	* COLON
```
The immediate reason for the failure is that the string `:four` matches `NODELIM` as well as the expected `:`, and, being longer, wins. However, NODELIM should not be in the list of tokens that the contextual parser can match in this state. My investigations suggest that the contextual lexer for the state that arises after the table 'key' is the initial follow set for the DELIM terminal, which includes all `data` regexps.

Interestingly, if `table_entry` is defined as `DELIM ":" data` (i.e. removing the `quoted_value` rule) parsing is successful.

BTW I initially ran across this in my Julia port of Lark (Lerche) based on Lark 0.7 so it is a high-fidelity copy!

**Comment by MegaIng** [comment]
I think this is a non fixable problem. I assume you include the `:` for a reason in you values, so that fix probably won't work. Another fix is to add a space after `:`, but this isn't really a bug in Lark that can be fixed.

**Comment by erezsh** [comment]
> Interestingly, if table_entry is defined as DELIM ":" data (i.e. removing the quoted_value rule) parsing is successful.

That is exactly what you should expect from LALR(1).

LALR is very predictive on shifts, but a less predictive on reduces, since it only detects that rules follow one another, but not *when* do they follow each other and when they don't (I can try a less hand-wavy explanation if you'd like).

You prevented the reduce of `quoted_value` and turned it instead into a shift of `DELIM`, which helped the parser by being more predictive.

In short, this isn't a bug in Lark, but a property of the LALR(1) algorithm.

---

On another matter, how is Lerche doing?

Are you finding it useful? Are you considering updating it with recent features?

**Comment by jamesrhester** [comment]
Thanks for the explanation. I was assuming that the contextual parser had been primed with a list of tokens that could legally follow for each state of the parser, but I see that that might not be known for a given state until a few shift/reduces later.

As for Lerche, I am using it for the [DrelTools.jl](https://github.com/jamesrhester/DrelTools.jl) project. It has been useful for demonstrating that the EBNF developed for the dREL language is correct, and to quickly reject any proposed enhancements that lead to ambiguity (and of course for actually transforming expressions in dREL to Julia). I am just now adding a Lerche grammar to [CrystalInfoFramework.jl ](https://github.com/jamesrhester/CrystalInfoFramework.jl) to provide a pure Julia alternative to a hand-written C parser. It is unfortunately quite slow (factor of 10 compared to C and maybe factor of 2-ish compared to Lark) so before adding in the latest Lark updates I should probably figure out why it is quite so sluggish and actually apply a few Julian optimisation techniques.
